### ISSUE TITLE
json_encode incorrectly accepts and encodes "inf" and "nan"

### ISSUE THREAD

**Issue by dbarnett** [vim/vim#654]
`json_encode` accepts special "inf" and "nan" float values and renders them literally into the encoded JSON:

``` vim
:echo json_encode(0.0 / 0.0)
'-nan'
```

Those values aren't in the JSON spec, and it must have been an oversight since those values can't be subsequently decoded:

``` vim
:call json_decode(json_encode(0.0 / 0.0))
```

```
E474: Invalid argument
```

**Comment by brammool** [comment]
David Barnett wrote:

> `json_encode` accepts special "inf" and "nan" float values and renders them literally into the encoded JSON:
> 
> ``` vim
> :echo json_encode(0.0 / 0.0)
> '-nan'
> ```
> 
> Those values aren't in the JSON spec, and it must have been an
> oversight since those values can't be subsequently decoded:
> 
> ``` vim
> :call json_decode(json_encode(0.0 / 0.0))
> ```
> 
> ```
> E474: Invalid argument
> ```

Thanks for reporting.  json_encode() should produce null, at least
that's when happens in Javascript.

For js_encode() we can pass on the values.

## 

When I look deep into your eyes, I see JPEG artifacts.
I can tell by the pixels that we're wrong for each other.  (xkcd)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
2016-02-23 23:27 GMT+03:00 Bram Moolenaar vim-dev-github@256bit.org:

> David Barnett wrote:
> 
> > `json_encode` accepts special "inf" and "nan" float values and renders
> > them literally into the encoded JSON:
> > 
> > ``` vim
> > :echo json_encode(0.0 / 0.0)
> > '-nan'
> > ```
> > 
> > Those values aren't in the JSON spec, and it must have been an
> > oversight since those values can't be subsequently decoded:
> > 
> > ``` vim
> > :call json_decode(json_encode(0.0 / 0.0))
> > ```
> > 
> > ```
> > E474: Invalid argument
> > ```
> 
> Thanks for reporting. json_encode() should produce null, at least
> that's when happens in Javascript.
> 
> For js_encode() we can pass on the values.

I checked how this is handled in other languages:
- Python and Ruby (Python standard library 3.4.3-r1 and 2.7.10-r1, Python
  demjson 2.2.4, Python simplejson 3.8.1, Ruby 2.0.0_p647-r1) encode infinity
  as `Infinity` and NaN as `NaN` and is able to parse the whole thing back. I
  guess they have an agreement
- Perl (http://search.cpan.org/dist/JSON/) has the same problem as ​Vim.
- Lua has way too many JSON parsers to test, none of which are standard.
  dkjson uses `null`.
- Haskell’s `aeson` uses `null`, though it again is not the only or the
  standard package. There are not _that_ much packages as with lua here, but
  I have only aeson because it is used by pandoc and yaml.

If that was YAML it would be .nan and .inf (with rather big number of
variations like .NaN, .Inf, .INF, .NAN, but not e.g. .InF: these are
variants hardcoded in the standard and not case-insensitiviness).

Thus I would prefer in order
1. Simply error out. This is better then suddenly changing value type
   because if there is infinity somewhere then actually it was expected to be
   a float and immediate error when dumping is better then error in code that
   tried to use `null` as a float later.
2. Use `null`. Standard in fact has no `Infinity` or `NaN`, so Ruby/Python
   variant can be rather unexpected.
3. Support Ruby and Python, maybe this will make it into the next standard.
   I also hope they are not alone.

**Comment by brammool** [comment]
Nikolay Pavlov wrote:

> 2016-02-23 23:27 GMT+03:00 Bram Moolenaar vim-dev-github@256bit.org:
> 
> > David Barnett wrote:
> > 
> > > `json_encode` accepts special "inf" and "nan" float values and renders
> > > them literally into the encoded JSON:
> > > 
> > > ``` vim
> > > :echo json_encode(0.0 / 0.0)
> > > '-nan'
> > > ```
> > > 
> > > Those values aren't in the JSON spec, and it must have been an
> > > oversight since those values can't be subsequently decoded:
> > > 
> > > ``` vim
> > > :call json_decode(json_encode(0.0 / 0.0))
> > > ```
> > > 
> > > ```
> > > E474: Invalid argument
> > > ```
> > 
> > Thanks for reporting. json_encode() should produce null, at least
> > that's when happens in Javascript.
> > 
> > For js_encode() we can pass on the values.
> 
> I checked how this is handled in other languages:
> - Python and Ruby (Python standard library 3.4.3-r1 and 2.7.10-r1, Python
>   demjson 2.2.4, Python simplejson 3.8.1, Ruby 2.0.0_p647-r1) encode infinity
>   as `Infinity` and NaN as `NaN` and is able to parse the whole thing back. I
>   guess they have an agreement
> - Perl (http://search.cpan.org/dist/JSON/) has the same problem as ​Vim.
> - Lua has way too many JSON parsers to test, none of which are standard.
>   dkjson uses `null`.
> - Haskell’s `aeson` uses `null`, though it again is not the only or the
>   standard package. There are not _that_ much packages as with lua here, but
>   I have only aeson because it is used by pandoc and yaml.
> 
> If that was YAML it would be .nan and .inf (with rather big number of
> variations like .NaN, .Inf, .INF, .NAN, but not e.g. .InF: these are
> variants hardcoded in the standard and not case-insensitiviness).
> 
> Thus I would prefer in order
> 1. Simply error out. This is better then suddenly changing value type
>    because if there is infinity somewhere then actually it was expected to be
>    a float and immediate error when dumping is better then error in code that
>    tried to use `null` as a float later.
> 2. Use `null`. Standard in fact has no `Infinity` or `NaN`, so Ruby/Python
>    variant can be rather unexpected.
> 3. Support Ruby and Python, maybe this will make it into the next standard.
>    I also hope they are not alone.

Thanks for looking into that.  I was following JavaScript, but I'm not
actually happy with it.  Turning NaN and Infinity into null is not very
useful.  It appears this is just done because the standard says NaN and
Infinity don't exist, which appears to be a shortcoming of the standard.

I think following Ruby and Python is most useful.  If the other side
does not understand NaN or Infinity then you'll get the error that your
first point suggests, only it's going to be at the receiver side.  In
practice this should work, the receiver probably will generate errors
for content it can't accept anyway.

Let's hope there will be a new standard.

## 

TIM:    Too late.
ARTHUR: What?
TIM:    There he is!
   [They all turn, and see a large white RABBIT lollop a few yards out of the
   cave.  Accompanied by terrifying chord and jarring metallic monster noise.]
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
That has been fixed and can be closed, right?

**Comment by lacygoill** [comment]
> That has been fixed and can be closed, right?

Yes, I think it was fixed by [7.4.1407](https://github.com/vim/vim/releases/tag/v7.4.1407).

Starting from this patch the first command returns `null`:

    :echo json_encode(0.0 / 0.0)
    null

And the second command does not raise any error anymore:

    :call json_decode(json_encode(0.0 / 0.0))

And starting from [7.4.1430](https://github.com/vim/vim/releases/tag/v7.4.1430), the first command returns `NaN`:

    :echo json_encode(0.0 / 0.0)
    NaN

**Comment by chrisbra** [comment]
thanks, closing.
